java/arcs/core/util/Result.kt

```diff
@@ -30,3 +30,28 @@ suspend fun <T> resultOfSuspend(block: suspend () -> T): Result<T> = try {
 } catch (e: Throwable) {
     Result.Err(e)
 }
+
```

Comment:
I'm just seeing this class for the first time, and the `resultOf` method looks a bit troubling to me. It swallows *all* exceptions in the block, which might include things that weren't intended (e.g. programmer errors like NullPointerException, IllegalArgumentException). With the addition of `getOrNull` it is now very easy to accidentally and silently swallow exceptions that should instead fail very loudly.

I have a couple of ideas:

1. Parameterise the Result type with an exception type, `Result<T, E>`, and then only catch the exception of type `E` in `resultOf`.
2. Don't accept a block, but instead provide a `resultOf(exception: Throwable)` function, which will force people to manually wrap their exceptions. With the Kotlin `return@blah` statement, it should be quite easy to escape out of deeply nested lambdas and return a failure, e.g. `return@forEach resultOf(NoSuchElementException("..."))`

I'd personally prefer option 2, and would suggest helper methods to let you write natural sounding things like `return ok(123)` or `return error(NoSuchElementException())`

Reply:
For some context, @jasonwyatt added this for his storage use cases. So, he might have more context.   I found the abstraction useful for my cases too. Hence, I started expanding it a bit. 

This class is modeled after https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/ 
The whole point of Kotlin's `Result` class is to be able to have referential transparency even in the presence of exceptions. The `resultOf` method is very similar to the [Result.mapCatching](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/map-catching.html) api. The `Result.Err` type would let the user of `resultOf` to inspect what exceptions were raised and act accordingly. 

Your next question would naturally be why don't we use the Kotlin's Result directly? :) Using `Result` as a return type is  flagged as an error by the Kotlin compiler.  See https://github.com/Kotlin/KEEP/blob/master/proposals/stdlib/result.md#limitations. As I understand, they want to expand or change the semantics of `Result` and don't want to deal with backward compatibility for this type yet. One of my guesses is that they might want add a monad like the safe call operator for `Result`.

Reply:
Thanks for the additional context.

I think it's the combination of catching Throwable mixed with getOrNull that worries me the most. Transforming exceptions is fine, as long as someone at the very end handles them all appropriately. Throwing the exception away and returning null throws away a lot of useful information, and I'm concerned that big bugs will be completely hidden.

Something I've seen in the past is an API where you pass in your expected exception type, and if the exception doesn't match it will throw. e.g. `getOrNull(NoSuchElementException::class)` does what it says if the exception is the right type, but if it's something else it will rethrow.

Reply:
Thought about this a bit and decided to abandon this in favor of a simpler class that does not involve exceptions: https://github.com/PolymerLabs/arcs/pull/5028